### DISCUSSION TITLE
[typescript] import { Map } from maplibre-gl gives error 'cannot find Map' in browser

### DISCUSSION THREAD

**Discussion by Skyaero42** [maplibre/maplibre-gl-js#3859] [Q&A]
Software:

- Visual Studio 2022
- Libman to load libraries
- maplibre-gl 4.1.1
- Firefox

I'm trying to reproduce [this](https://maplibre.org/maplibre-gl-js/docs/) example in Typescript using [this](https://maplibre.org/maplibre-gl-js/docs/API/) import example.

mymap.ts:
```
import { Map } from 'maplibre-gl';

new Map({
    container: 'map',
    style: 'https://demotiles.maplibre.org/style.json',
    center: [-74.5, 40],
    zoom: 9
});
```

index.htm:
```
    <script type="importmap" asp-append-version="true">
        {
          "imports": {
            "maplibre-gl": "https://unpkg.com/maplibre-gl/dist/maplibre-gl.js"
          }
        }
    </script>
    <script src="~/js-test/mymap.min.js" type="module" asp-append-version="true"></script>
```

Typescript compiles, but the browser gives the error *import not found: Map*

I've also tried (while using `new maplibregl.Map()` instead of `new Map()`):
```
// import not found: default
import maplibregl from 'maplibre-gl';

// import not found: maplibregl
import { maplibregl } from 'maplibre-gl'

// maplibregl.Map is not a constructor
import * as maplibregl from 'maplibre-gl'
```

Is something going wrong with the exports?

**Comment by HarelM** [comment]
How do you compile an html file as typescript? This is strange...
Please provide stackblitz/jsbin reproduction.

**Comment by Skyaero42** [comment]
> How do you compile an html file as typescript? This is strange... Please provide stackblitz/jsbin reproduction.

I compile the mymap.ts file to mymap.min.js and have that javascript file loaded by the html page.

**Comment by neodescis** [comment]
I think most folks also use some sort of bundler, but it seems like this should work. I've never used an importmap though. It might be somewhat complicated by "Map" already existing though. Have you tried aliasing the import?

`import { Map as MapLibre } from 'maplibre-gl'`

**Comment by Skyaero42** [comment]
importmap is needed to be able to use a CDN, otherwise the import always looks for a local file.

I've simplified the problem, because it looks like it has nothing to do with typescript.

```html
<html>
<head>
    <script type="importmap">
        {
            "imports": {
                "maplibre-gl": "https://unpkg.com/maplibre-gl@4.1.1/dist/maplibre-gl.js"
            }
        }
    </script>
    <script type="module">
        import { Map } from 'maplibre-gl';

        new Map({
            container: 'map',
            style: 'https://demotiles.maplibre.org/style.json', 
            center: [-74.5, 40],
            zoom: 9
        });
    </script>
</head>
<body>
    <section id="map">
    </section>
</body>
</html>
```

This should work, but doesn't as it cannot find 'Map'. An alias doesn't solve it either.

```
import * as M from 'maplibre-gl'
```
results in a module M that exists, but that module does not contain M.Map() or M.default.Map().

I think something is going wrong with exports in the maplibre-gl library.

**Comment by HarelM** [comment]
The following will work due to how maplibre package is built:
```html
<html>
<head>
    <script type="importmap">
        {
            "imports": {
                "maplibre-gl": "https://unpkg.com/maplibre-gl@4.1.1/dist/maplibre-gl.js"
            }
        }
    </script>
    <script type="module">
        import 'maplibre-gl';

        new maplibregl.Map({
            container: 'map',
            style: 'https://demotiles.maplibre.org/style.json', 
            center: [-74.5, 40],
            zoom: 9
        });
    </script>
</head>
<body>
    <section id="map">
    </section>
</body>
</html>
```

**Comment by HarelM** [comment]
See also:
- #1595
- #977

**Comment by creativetags** [comment]
For anyone else struggling to get this to work I had to do:
`import maplibregl from 'maplibre-gl'`
